### ISSUE TITLE
Bug: Exceptions not handled correctly

### ISSUE THREAD

**Issue by jkpubsrc** [agronholm/typeguard#34]
Situation:

I've a function here with return value constraint `List[xyz]`. This function throws an Exception for some reason.

Encountered Behavior:

`typeguard` complains that the return value of this function is `None`.

Expected Behavior:

The exception should be let through. `typeguard` should not consume that exception.

Details:

I conclude that there is a bug because if I modify the return value constraint from `List[xyz]` to `Union[List[xyz], None]` everything works fine. In that modified case the Exception from within that function is not swallowed but raised correctly. To my understanding exceptions should never be swallowed if they are raised. Return type checking should be performed only if no exception is raised. Exceptions should have priority here: No data is not returned anyway if an exception is raised within a function. So no return type checking should be performed if `typeguard` encounters an exception.

I guess some kind of very minor change within the code could fix that problem? Despite of such a minor glitch `typeguard` works quite well!

**Comment by agronholm** [comment]
Typeguard offers three modes of operation. Which one are you using?

**Comment by agronholm** [comment]
Also please add a reproducible test case.

**Comment by jkpubsrc** [comment]
Can't reproduce it. The test case I tried to implement for you right now works just fine. Which proofs things are not that easy as I thought.

And right now I can't reproduce the situation in my large program where I encountered this problem: I made modifications to the data my program uses so I've a completely different situation now. But I remember: It was clearly reproducible. It made me wonder quite a lot. After commenting out your module and the wrapper code activating typeguard everything worked as expected: The exception got thrown as expected.

What keeps me thinking: I'm doing some very very fancy things with catching exceptions and dynamically loading classes and invoking methods on them. (That's why I wasn't able to provide a test case right from the start.) Maybe for some reasons this interferes somehow with your module though right now I have no idea how that could be.

As the isolated test case I tried to create for you worked just fine I don't see where you, me or s.o. else could start right away fixing this problem. My recommendation: Close this issue. If I encounter this bug again I'll try to reopen this issue if possible or file a new bug report instead. Hopefully than with more information.
